James 2156

Fredmomo opened the pull request.

mbaechler committed:
LOCAL ONLY max C* mem

mbaechler committed:
JAMES-2544 some tests probably need more time than expected

mbaechler committed:
JAMES-2630 Port CassandraMessageMapper to Reactor

aduprat committed:
JAMES-2620 Built artifacts should be installed to prevent downloading them from maven repository

mbaechler committed:
JAMES-2544 implement RabbitMQ dequeue with reactor-rabbitmq

aduprat committed:
JAMES-2622 Introduce env variable for java options in dockerfiles

mbaechler committed:
JAMES-2544 remove unused RabbitChannelPoolImpl

blackheaven committed:
JAMES-2630 Migrate CassandraAsyncExecutor.executeReturnExists consumers to Reactor

blackheaven committed:
JAMES-2630 Migrate CassandraAsyncExecutor consumers to Reactor for event-sourcing-event-store-cassandra

blackheaven committed:
JAMES-2630 Migrate CassandraAsyncExecutor.executeReturnExists consumers to Reactor contd

chibenwa committed:
MAILBOX-364 EventFactory should expose a builder for MailboxAdded event

chibenwa committed:
MAILBOX-364 EventFactory should expose a builder for FlagsUpdated event

chibenwa committed:
MAILBOX-364 EventFactory should expose a builder for MailboxDeletion event

chibenwa committed:
MAILBOX-364 EventFactory should expose a builder for Added event

chibenwa committed:
MAILBOX-364 EventFactory should expose a builder for Expunged event

chibenwa committed:
MAILBOX-364 EventFactory should expose a builder for MailboxAclUpdated event

chibenwa committed:
MAILBOX-364 EventFactory should expose a builder for MailboxRenamed event

chibenwa committed:
MAILBOX-364 EventFactory should rather link LessageMoveEvent.Builder

chibenwa committed:
MAILBOX-364 EventFactory methods can be statics

chibenwa committed:
MAILBOX-359 Remove MailboxEventDispatcher::ofMailboxListener

chibenwa committed:
MAILBOX-359 MaildirMailboxMapper should return the mailboxId on persisted mailboxes

mbaechler committed:
JAMES-2544 ensure rabbitFlux is closed

mbaechler committed:
JAMES-2544 upgrade rabbitmq client library

mbaechler committed:
JAMES-2544 migrate SendMDNMethodTest to junit 5

chibenwa committed:
MAILBOX-359 Factorize tests for MessageUid ser-deserialization

chibenwa committed:
MAILBOX-359 Factorize tests for MessageMetaData ser-deserialization

chibenwa committed:
MAILBOX-359 Re-arrange QuotaUsageUpdatedEventSerializationTest

WithTime success case is strictly equivalent to the success case

chibenwa committed:
MAILBOX-359 Factorize tests for UpdatedFlags ser-deserialization

chibenwa committed:
MAILBOX-359 Factorize tests for ACLDiffTest ser-deserialization

chibenwa committed:
MAILBOX-359 Use a fixture for the serializers

chibenwa committed:
MAILBOX-359 Factorize tests for Quota ser-deserialization

chibenwa committed:
MAILBOX-366 Create an EventId POJO

chibenwa committed:
MAILBOX-366 Events should carry an EventId

chibenwa committed:
MAILBOX-359 Switch Scala Event trait back to private

This needs to be done by wrapping all references to this trait as private

chibenwa committed:
MAILBOX-359 Use val to avoid too much class instanciations

hoangdat committed:
MAILBOX-369 Add EventBusId POJOs

hoangdat committed:
MAILBOX-369 Mark IdleMailboxListener as ASYNCHRONOUS

chibenwa committed:
MAILBOX-364 MailboxEvent builders should use staged builders

We can leverage stage sharing through generics. Note that also uncouple stages from
each others (to define stage 1, I don't need to know stage 2, only the builder method defining the
stage chain needs that) which is a very nice benefit!

Note that alternative choices are implemented using default methods.

chibenwa committed:
MAILBOX-364 Builder for QuotaUsageUpdatedEvent

chibenwa committed:
JAMES-2636 Wildcard definition needs to be covered by MappingSource

This constant was only used by tests and made the API confusing.
Refactoring this was easy.

chibenwa committed:
JAMES-2636 Reorder RecipientRewriteTable API order

Classes go first, then constants, then methods. Note that I inlined a
short method that was not related to the API. This avoids poluting the API
with useless methods

chibenwa committed:
JAMES-2636 Enhence RRT API getMapping methods

The difference between them was subtil (resolved/unresolved) and prooved
to be confusing. Hence I propose a better method naming for solving this
issue.

chibenwa committed:
JAMES-2636 Remove serial UID from ErrorMappingException

Arsnael committed:
JAMES-2636 Add 'Alias' mapping type and handle rewrites for it

Arsnael committed:
JAMES-2636 getUserDomainMappings should return empty object consistently when there is no mappings

Arsnael committed:
JAMES-2636 refactoring some mapping tests

trantienduchn committed:
Adding LICENSE for project

hoangdat committed:
MAILBOX-367 Factorise RabbitMQ constants

Arsnael committed:
JAMES-2636 Refactoring and automating some mapping methods

hoangdat committed:
MAILBOX-367 EventBusContract should use real event pojo

To avoid scala serialization error happens on mock event while running
rabbitmq eventbus

hoangdat committed:
MAILBOX-367 Add mockito timeouts in EventBus contracts

This allow asynchronous EventBus implementations
EventBusContract should also use verify no interaction in time window

hoangdat committed:
MAILBOX-367 RabbitMQEvenBus should publish events

chibenwa committed:
JAMES-2641 Remove MailboxEventDispatcher tests

This class is going to be removed later in that PR

chibenwa committed:
JAMES-2641 AbstractMessageIdManagerSideEffectTest should not rely on mock assertions

This makes change harder to test and depends heavily on the underlying implementation

chibenwa committed:
JAMES-2641 Remove MailboxEventDispatcher extra methods

Callers should use directly the builders and call the `event` method

This transform the MailboxEventDispatcher into a useless wrapper around
the delegatingMailboxListener that is now easier to remove.

Arsnael committed:
JAMES-2636 Refactoring AbstractRecipientRewriteTableTest test class to enhance clarity

chibenwa committed:
JAMES-2641 Remove MailboxEventDispatcher

Let's emmit on the Delegating listener

chibenwa committed:
JAMES-2631 ReIndent InMemoryHostSystem

Arsnael committed:
JAMES-2636 add and remove alias mapping for RecipientRewriteTable and cucumber tests

chibenwa committed:
MAILBOX-359 Remove a unused method

chibenwa committed:
MAILBOX-359 Improving QuotaScope coverage

chibenwa committed:
MAILBOX-359 ReIndent QuotaUsageUpdatedEventSerializationTest

chibenwa committed:
JAMES-2641 Re-organize mailbox manager fields

chibenwa committed:
JAMES-2641 Make StoreMailboxManager::idGenerator final

chibenwa committed:
JAMES-2641 Make StoreMailboxManager::batchSizes final

To do this we need to let the injection system providing the built value.

Note that we inject a more generic "MailboxManagerConfiguration" class in order to make
future parameters addition easier in the future.

Note that Sprint (& memory) did not configure the batch size. Thus to reduce diff size the constructors directly provide default value as a temporary solution.

chibenwa committed:
JAMES-2641 Make StoreMailboxManager::Authorizator field final

chibenwa committed:
JAMES-2641 Extract Session management from StoreMailboxManager

chibenwa committed:
JAMES-2641 DefaultUserQuotaRootResolver should rely on SessionProvider

This allow uncoupling it from the mailbox manager and breaks a dependency loop that prevented final allocation

hoangdat committed:
MAILBOX-368 Separating Group and Key Contract from EventBusContract

hoangdat committed:
MAILBOX-368 EventBusContract should enforce basic error handling checks

- Add one test when failed before, consumer is still able to handle subsequent events
- A failing listener should not prevent subsequent listeners execution

hoangdat committed:
MAILBOX-368 RabbitMQEventBus group registration

hoangdat committed:
MAILBOX-368 Add Multiple EventBus tests for group registration

blackheaven committed:
JAMES-2630 Upgrade Reactor

blackheaven committed:
JAMES-2630 Add Functional utils

blackheaven committed:
JAMES-2630 Add Reactor utils

Arsnael committed:
JAMES-2637 add PUT route for aliases routes

chibenwa committed:
JAMES-2641 BatchSizes can be missing

hoangdat committed:
MAILBOX-368 Add EventBus Concurrent test for group registration

chibenwa committed:
JAMES-2641 Fix QuotaMessageManagerTest

The QuotaUpdater was registered 2 times

chibenwa committed:
JAMES-2641 Stop lazy initialization of Quota components in MailboxManager

chibenwa committed:
JAMES-2641 MailboxAnnotationListener should depend on SessionProvider

Again this breaks a dependency loop

trantienduchn committed:
MAILBOX-373 EventDeadLetters API

trantienduchn committed:
MAILBOX-373 EventDeadLetters Contract & Memory impl

chibenwa committed:
JAMES-2641 Search indexes should not be initialized in StoreMailboxManager::init()

blackheaven committed:
JAMES-2630 Migrate CassandraAsyncExecutor.executeReturnApplied consumers to Reactor

hoangdat committed:
MAILBOX-367 Simplify RabbitMQManagement creations

chibenwa committed:
JAMES-2641 Search indexes should rely on SessionProvider

This again avoids a circular dependency

chibenwa committed:
JAMES-2641 Simplify slightly InMemoryIntegrationResources

chibenwa committed:
JAMES-2641 Stop lazy initialization of Quota components in MailboxManager

chibenwa committed:
JAMES-2641 Finally remove StoreMailboxManager::init()

It's last responsibility, registrating listeners, needs to be performed explicitly

That way StoreMailboxListener no more need knowledge about components composing it.

chibenwa committed:
JAMES-2641 Correct spring logic

We need to register listeners that was previously registered

chibenwa committed:
MAILBOX-377 Disable SpamAssassinInvokerTest::scanMailShouldMarkAsSpamWhenKnownAsSpam

hoangdat committed:
MAILBOX-367 Moving RabbitMQManagement API to rabbitmq-backend project

- Using this API in RabbitMQ EventBus

Arsnael committed:
JAMES-2637 refactoring group and forward routes

trantienduchn committed:
MAILBOX-372 InVmEventDelivery returns success even when exception tests

trantienduchn committed:
MAILBOX-372 InVmEventDelivery exponential backoff

- InVmEventDelivery returns error when delivery fails, but always
returns success when dispatching
- Update ErrorHandlingContract to EventBusContract, testing max retries
and exponential backoff delay

hoangdat committed:
MAILBOX-367 Expose Exchange APIs to RabbitMQManagementAPI

chibenwa committed:
JAMES-2641 Delete DefaultDelegatingMailboxListenerTest

Arsnael committed:
JAMES-2636 refactoring mapping types

chibenwa committed:
JAMES-2641 Introduce GroupMailboxListener

That is mailboxListener carrying their default execution group around.

chibenwa committed:
JAMES-2641 SetMessagesMethodTest should be using EventCollector

Rather than implementing its own

chibenwa committed:
JAMES-2641 MailboxManager should no more implement MailboxListener support

It should rather expose proxy method for the event bus.

Note that these proxy method can be temporarily implemented on top of the Delegating Listener

hoangdat committed:
MAILBOX-367 RabbitMQEventBus exchange, queue life cycle tests

chibenwa committed:
JAMES-2641 MessageMoves should return impactedMailboxIds

This is needed to get the registration keys to dispatch related events

chibenwa committed:
JAMES-2641 Drop Delegating mailbox listener and rather use the EventBus

chibenwa committed:
JAMES-2641 Remove MailboxListenerRegistry

chibenwa committed:
JAMES-2641 Remove now useless ListenerType

chibenwa committed:
JAMES-2641 MailboxManager should not expose group registration

hoangdat committed:
MAILBOX-367 Disable using Nio in RabbitMQ ConnectionFactory

- While using Nio, RabbitMQEventBusTest running is not stable.
Some tests sometimes be hang for a long times (ie: 9 minutes).

chibenwa committed:
JAMES-2641 DefaultImapProcessorFactory have many unused fields/methods

Only 2 static methods are used

chibenwa committed:
JAMES-2641 DefaultProcessorChain should not use final variable/parameters

chibenwa committed:
JAMES-2641 Re-order fields of AbstractSelectionProcessor

chibenwa committed:
JAMES-2641 IMAP components should use the EventBus directly

This enables the MailboxManager to no more be an EventBus proxy

chibenwa committed:
JAMES-2641 IMAP components should use the EventBus directly

This enables the MailboxManager to no more be an EventBus proxy

chibenwa committed:
JAMES-2641 EventBus caller should block to await dispatching to be done

hoangdat committed:
MAILBOX-371 Registration should have a deserialization factory

Arsnael committed:
JAMES-2637 add rest of the routes for alias routes

Arsnael committed:
JAMES-2637 refactoring forward and group routes

hoangdat committed:
MAILBOX-371 Conversion between RoutingKeys And RegistrationKeys

Arsnael committed:
JAMES-2637 fix alias mappings for PUT alias route

trantienduchn committed:
MAILBOX-372 suppress InVmEventDelivery errors in DelegateMailBoxListener

Before that InVmEventDelivery was already suppressed, it always returns
success Mono.

DelegateMailBoxListener will soon be removed. So keeping its behavior till
it get removed.

hoangdat committed:
MAILBOX-371 RabbitMQEventBus impl for key registrations

chibenwa committed:
JAMES-2641 Enhance ListeningMessageSearchIndex logs

chibenwa committed:
JAMES-2641 Ship knowledge about serializability along in MessageId

JPA integration test issues some failed logs about "fake" messageId serializing
that should have been prevented.

chibenwa committed:
JAMES-2641 Enable Annotations in JPA Guice

trantienduchn committed:
MAILBOX-372 update tests affected by retry mechanism

using Mockito mock invocations (call real method, throwing...)
doesn't make test defeats RabbitMQ asynchronous arbitrary dispatching
with retry backoff.

Define new dedicated listener class for this purpose.

chibenwa committed:
JAMES-2641 Avoid retrieving mailbox information upon renames

This is simply not needed and causes the JPA + Lucene product to fail

chibenwa committed:
JAMES-2641 Allow specifying group of registered listeners

This allow duplicate listeners to be registered, a feature used by over-quota mailing
to specify several thresholds.

hoangdat committed:
MAILBOX-371 Remove unnecessary test for lifecycle

This test is in SingleLifeCycleTest, There is another equivalent test
with multiple event bus. Remove this because it is impossible to passed
when RabbitMQEventBusTest also implements MultipleLifeCycleTest, then in
setup phrase, two eventbus are up, that means, stopping one eventbus
doesn't make queue registration for key deleted. (Because there is another
eventbus is consuming this queue!)

hoangdat committed:
MAILBOX-371 Wrapping RabbitMQExtension to avoid start and stop several rabbit mq docker

hoangdat committed:
MAILBOX-371 Add Concurrent Test for RabbitMQEventBus with Key Registration

hoangdat committed:
MAILBOX-371 Fix-up for LifeCycleTest more stable

Arsnael committed:
JAMES-2642 alias webadmin integration test

chibenwa committed:
MAILBOX-377 remove unused method in EventBus ErrorHandling contract

chibenwa committed:
JAMES-2641 DefaultEventBus should not bind specific listeners

chibenwa committed:
MAILBOX-377 Slightly improve InVMEventDelivery

chibenwa committed:
MAILBOX-377 Fix asynchronous tests for In VM event delivery

chibenwa committed:
MAILBOX-377 Increase some too short timeouts in RabbitMQ event bus tests

Slow computer may need more than 5 seconds to handle 10.000 deliveries...

Arsnael committed:
JAMES-2637 add asMailAddressString method to MappingSource

Arsnael committed:
JAMES-2637 Rely on asMailAddressString for MappingSource serialization

Yolanda-Tang committed:
JAMES-2156 add attachment mapper

trantienduchn committed:
MAILBOX-372 RetryBackoff POJO

trantienduchn committed:
MAILBOX-372 generating wait delay in WaitDelayGenerator

Arsnael committed:
JAMES-2637 Strong typing for AliasSourcesResponse

trantienduchn committed:
MAILBOX-372 handle retrying for event group consumers

Arsnael committed:
JAMES-2637 getAliasSources should use rrt.listSources

Arsnael committed:
JAMES-2637 add tests for getSourcesForType

Arsnael committed:
JAMES-2637 add getMappingsForType method with tests

Arsnael committed:
JAMES-2637 return streams for getSourcesForType and getMappingsForType methods

trantienduchn committed:
MAILBOX-372 ErrorHandling is applied for only group registered listeners

chibenwa committed:
JAMES-2641 Harden MailboxManagerTest for events and do some fix

We need to also ensure the right dispatch key is being used...

Arsnael committed:
JAMES-2637 add test to MappingSource to be sure it matches the bean contract

hoangdat committed:
MAILBOX-369 Using EventBusId to skip synchronous local registered listener

chibenwa committed:
MAILBOX-372 Allow MailboxListeners to throw exception

chibenwa committed:
MAILBOX-372 ElasticSearchQuotaMailboxListener should not catch exceptions

chibenwa committed:
MAILBOX-372 ListeningCurrentQuotaUpdater should not catch exceptions

chibenwa committed:
MAILBOX-372 ListeningMessageSearchIndex should not catch exceptions

chibenwa committed:
MAILBOX-372 MailboxAnnotationListener should not catch exceptions

chibenwa committed:
MAILBOX-372 PropagateLookupRightListener should not catch exceptions

chibenwa committed:
MAILBOX-372 QuotaThresholdCrossingListener should not catch exceptions

chibenwa committed:
MAILBOX-372 SpamAssassinListener should not catch exceptions

Arsnael committed:
JAMES-2642 Add basic alias mapping integration tests

Arsnael committed:
JAMES-2642 add a few more alias mapping integration tests

Arsnael committed:
JAMES-2642 Refactor and cleanup code for alias, group and forward integration tests

trantienduchn committed:
MAILBOX-372 make configuration public

trantienduchn committed:
MAILBOX-372 Use RetryBackoffConfiguration in InVmEventDelivery

blackheaven committed:
JAMES-2644 Fix the Content-Type parsing fragility

chibenwa committed:
MAILBOX-374 Bind RabbitMQEventBus module in RabbitMQ related product

chibenwa committed:
JAMES-2641 Add a missing import after rebase

Arsnael committed:
JAMES-2645 Refactor cassandra RRTs by adding a proper DAO class with basic tests

trantienduchn committed:
MAILBOX-372 RabbitMQEventBus should receive events didn't fully processed

chibenwa committed:
MAILBOX-374 RabbitMQ should support generic groups

This bug caused only one of the QuotaMailing listeners to be triggered

trantienduchn committed:
MAILBOX-376 Plug EventDeadLetter into InVmEventBus

chibenwa committed:
MAILBOX-374 Checking registration queues deletion should filter out group queues

chibenwa committed:
MAILBOX-371 Test out of order registrations

chibenwa committed:
MAILBOX-372 Test dispatching from the listeners

This behaviour corresponds to ListeningCurrentQuotaUpdater

We demonstrate here that InVMEventDelivery do not support dispatch while retrying.

hoangdat committed:
MAILBOX-369 Using RabbitMQExtension directly for stabling RabbitMQEventBus
concurrent test

Arsnael committed:
JAMES-2646 list sources cassandra table, DAO and tests

chibenwa committed:
MAILBOX-374 Configure SpamAssassin only when required

This avoid the retry to happen.

chibenwa committed:
MAILBOX-374 QuotaRatio Index and Aliases should not include capital letters

This is not supported by ElasticSearch and leads to many EventBus retries

mbaechler committed:
JAMES-2630 Improve ConcurrentTestRunner reporting

mbaechler committed:
JAMES-2630 Improve Cassandra logging

blackheaven committed:
JAMES-2630 Set CassandraAsyncExecutor.executeReactor to defer to ensure composability

trantienduchn committed:
MAILBOX-369 Isolate concurrent tests to a nested class

Therefore they could not affect to other tests cases because concurrent
tests are running on another docker rabbitmq. Experiments show that
concurrent tests implicitly impact to other tests correctness, because they
fire a lot of events without consuming them, then later, other tests starting
consume events are fired from concurrent tests, and make assertions fail

To ensure docker rabbitmq is cleaned up after each test are quite difficult
to archive because some queues, exchanges are durable, even start & stop
docker doesn't make them abandon there messages. This could be considered as
a workaround for this issue. Creating new docker and start it in every test
is expensive and leads to minutes to finish this test class.

trantienduchn committed:
MAILBOX-376 WorkQueueName need to carry a Group instance

trantienduchn committed:
MAILBOX-376 Plug DeadLetters into RabbitMQEventBus

trantienduchn committed:
JAMES-2618 Missing SchemaVersion Module causes tests fail to setup

Arsnael committed:
JAMES-2647 Cassandra migration task for mapping sources

blackheaven committed:
JAMES-2578 Fix warning using Mailet new API

chibenwa committed:
MAILBOX-374 Fix test class name

Test was duplicated in it

chibenwa committed:
MAILBOX-374 IMAP MPT tests for rabbitmq reactor

These tests serves as integretion tests for IMAP and thus are important.

chibenwa committed:
MAILBOX-374 Fix a bug between group registration and key registration

In some cases groups were not dispatched when a key was registered/unregistered.

chibenwa committed:
MAILBOX-374 Slightly improve ListeningCurrentQuotaUpdater code quality

chibenwa committed:
MAILBOX-374 Use InMemoryIntegrationResources to ease RabbitMQ hostSystem instantiation

chibenwa committed:
MAILBOX-374 DockerRabbitMQ should be able to provide a connection factory

This avoids code duplication between JUNIT 4 & JUNIT 5

Arsnael committed:
JAMES-2648 cassandra list sources support implementation

chibenwa committed:
MAILBOX-374 reuse forks for RabbitMQ MPT testing

This allow to perform all the tests on a single RabbitMQ container

chibenwa committed:
MAILBOX-376 Re-indent event bus

chibenwa committed:
MAILBOX-376 Refactor GroupConsumerRetry

 - Remove unneeded RetryPublisher inner class
 - Avoid some not needed mono wrapping - which was leading to confusing method names

chibenwa committed:
MAILBOX-376 QueueName is only required when initializing group retry

chibenwa committed:
MAILBOX-376 Solve various IntelliJ warnings for RabbitMQEventBus

Including:

 - Visibility that could be more restricted
 - Lambdas that could be replaced by method reference
 - Some constants duplication

chibenwa committed:
MAILBOX-376 Structured logging for groups retries

chibenwa committed:
MAILBOX-376 Do not lose events "forever" when publishing retries fails

Keeping them in dead-letter feels better...

chibenwa committed:
MAILBOX-376 Structured logging for group listener execution

chibenwa committed:
MAILBOX-376 Structured logging for event dispatching

chibenwa committed:
MAILBOX-376 Structured logging for key registrations

chibenwa committed:
MAILBOX-376 Structured logging for local mailbox listener execution

chibenwa committed:
MAILBOX-376 Structured logging for InVmEventDelivery

chibenwa committed:
MAILBOX-376 Metrics for RabbitMQ event-bus listener execution time

chibenwa committed:
MAILBOX-364 Move event related classes in the event package of mailbox-api

chibenwa committed:
MAILBOX-374 Write a test about error recovery

chibenwa committed:
MAILBOX-374 RabbitMQConnectionFactory should rely on reactor for retries

chibenwa committed:
MAILBOX-374 Our connection mono should cache the connection as long as it is open

We should regenerate the connection mono when the connection had been close,
for instance due to error.

A custom callable had been written as I miss a `Mono<T>.cacheWhen(Predicate<T>)`

chibenwa committed:
MAILBOX-374 Specify unit for minDelays upon connection retries

chibenwa committed:
MAILBOX-374 Turn off exclusive - auto-deletes queues

In the face of network errors, we can not achieve a 'safety' and 'liveness' property.

 - 'liveness': A James node that is not failmed always have its Key registration queue
 - 'safety': The "key registration queue" gets deleted when there is not more consumer.

If the network partition exceeded the configured heartbeat, then the queue is going to be deleted and James is going to fail (as its queue disapeared).

We choose the safer option of not having RabbitMQ responsible of automatic queue deletion. The admin is going to be responsible of queue cleanup after crash.

chibenwa committed:
MAILBOX-374 Also test EventBus behaviour when confronted with RabbitMQ restarts

chibenwa committed:
MAILBOX-374 EventBusTestFixture should rely on java.time.Duration

chibenwa committed:
MAILBOX-374 Avoid a race condition upon connection recovery

This will lead to multiple connection opening, waisting resources

chibenwa commented on mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/JPAMailboxManager.java:
```diff
@@ -46,7 +46,8 @@
         MailboxCapabilities.Move,
-        MailboxCapabilities.Annotation);
+        MailboxCapabilities.Annotation,
+            MailboxCapabilities.Attachment);
```
Indent `MailboxCapabilities.Attachment` at the same level ^^

chibenwa commented on mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/model/openjpa/AbstractJPAMailboxMessage.java:
```diff
@@ -22,10 +22,8 @@
-import java.util.Comparator;
-import java.util.Date;
-import java.util.List;
+import java.util.*;
```
Wildcard spotted

chibenwa reviewed (commented):
Minor comments. Thanks for continuing this work!

chibenwa committed:
Update changelog: 25/01/2019

Arsnael committed:
JAMES-2649 Add a module for alias sources Cassandra projection consistency and tests

trantienduchn commented:
Hello @ReturnHttp402, the error log
```
org.apache.openjpa.persistence.OptimisticLockException: Attempted to attach deleted instance type "class org.apache.james.mailbox.jpa.mail.model.JPAMailbox" with oid
```
came from `@GeneratedValue` in `JPAMailbox`. In case storing new JPAMailbox, as I understand, JPA require the control to generate `mailboxId` but the code already set `mailboxId` in side the method `JPAUidProvider.lockedNextUid()`

I fear that removing `@GeneratedValue` may breaks james correctness in some where else. I'm investigating this issue. Seems like it's not easy to resolve

blackheaven committed:
JAMES-2652 Add Overflow test for POP3 RETR command

Fredmomo commented:
> Hello @ReturnHttp402, the error log
> 
> ```
> org.apache.openjpa.persistence.OptimisticLockException: Attempted to attach deleted instance type "class org.apache.james.mailbox.jpa.mail.model.JPAMailbox" with oid
> ```
> came from `@GeneratedValue` in `JPAMailbox`. In case storing new JPAMailbox, as I understand, JPA require the control to generate `mailboxId` but the code already set `mailboxId` in side the method `JPAUidProvider.lockedNextUid()`
> 
> I fear that removing `@GeneratedValue` may breaks james correctness in some where else. I'm investigating this issue. Seems like it's not easy to resolve

Hello @trantienduchn The default generated strategy is auto, if we change it to identity, according to OpenJPA's documentation the identity field should be added when inserting into table which may release the work of setting id manually in the code, and according to common JPA implementation, the id field would not be set after `persist()` into persist context, rather, it would be set after flushing into db using strategy, so it'd be very confused why the strategy does not work when persist/merge new entity of `JPAMailbox` into db, with throwing `CANNOT accept NULL for MAILBOX_ID` error.

trantienduchn commented:
I've just check your test. Here is a part of error logs I got, and I think that is what you also get when running the test:
```
Caused by: <openjpa-3.0.0-r422266:1833209 fatal general error> org.apache.openjpa.persistence.PersistenceException: Column 'MAILBOX_ID'  cannot accept a NULL value. {prepstmnt 1777381620 
INSERT INTO APP.JAMES_MAIL (MAILBOX_ID, MAIL_UID, MAIL_IS_ANSWERED, 
        MAIL_ATTACHMENT_ID, MAIL_BODY_START_OCTET, 
        MAIL_CONTENT_OCTETS_COUNT, MAIL_IS_DELETED, MAIL_IS_DRAFT, 
        MAIL_IS_FLAGGED, MAIL_DATE, MAIL_MIME_TYPE, MAIL_MODSEQ, 
        MAIL_IS_RECENT, MAIL_IS_SEEN, MAIL_MIME_SUBTYPE, 
        MAIL_TEXTUAL_LINE_COUNT, MAIL_BYTES, HEADER_BYTES) 
    VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?) 
```
So, in JPA, to add a message to a mailbox:
1. enrichMessage
2. save message with the mailbox

First step `enrichMessage` triggers mailbox creation when the mailbox is not existed. This happens inside `JPAUidProvider.lockedNextUid()`. you save a mailbox and let the JPA decides where the ID is generated. With `@GeneratedValue`, ID is generated by the database. Notice that, now, in `JPAUidProvider.lockedNextUid()`, `Mailbox mailbox` and `JPAMailbox m` have totally different IDs, `mailbox` has ID set by the code, `m` has ID generated by database.

Go to further steps, you `save message with the mailbox` (JPAMessageMapper.save()) but at that step the code get ID from the `mailbox` and then getting corresponding JPAMailbox from the ID of `mailbox` :
```java
            JPAId mailboxId = (JPAId) mailbox.getMailboxId();  // this mailboxId is not the ID saved in database
            JPAMailbox currentMailbox = getEntityManager().find(JPAMailbox.class, mailboxId.getRawId());  // then this statement should return null result. That leads to your query error
```

There is a workaround for this issue by reset ID of `mailbox` in `JPAUidProvider.lockedNextUid()` to the ID saved into database:
```java
            manager = factory.createEntityManager();
            manager.getTransaction().begin();
            JPAId mailboxId = (JPAId) mailbox.getMailboxId();
            JPAMailbox m = Optional.ofNullable(manager.find(JPAMailbox.class, mailboxId.getRawId()))
                    .orElse(new JPAMailbox(mailbox));
//            m.setMailboxId(mailboxId);
            m = manager.merge(m);
            long uid = m.consumeUid();
            mailbox.setMailboxId(m.getMailboxId());  // this line to reset mailbox id
            manager.getTransaction().commit();  
```
Then I think you will have a good mailbox id to use in `save` step

mbaechler committed:
JAMES-2630 change blobs-api to use Mono

mbaechler committed:
JAMES-2630 propagate reactor objects following blobs-api changes

mbaechler committed:
JAMES-2630 remove last calls to Mono::toFuture

mbaechler committed:
JAMES-2630 remove last calls to Mono::fromFuture

mbaechler committed:
JAMES-2630 fix ordering of message attachment issue

mbaechler committed:
Merge branch 'fix-master-reactor'

mbaechler committed:
Revert "LOCAL ONLY max C* mem"

This reverts commit a8142407c8e738ed984e2ea8077a41e49220afe9.

chibenwa committed:
JAMES-2654 Exclude logback dependency from CLI

chibenwa committed:
JAMES-2654 SMTP protocol module should depend on logback-classic only in tests

chibenwa committed:
JAMES-2655 Simplify JPA RRT query logic

Previous queries were complex and were syntactically incorrect when run against
MariaDB as reported by Marc Chamberlin.

This lead to portability issue.

Note that simplifying queries is an attempt to solve this.

chibenwa committed:
JAMES-2655 Strong typing for JPA RRT getMapping

chibenwa committed:
JAMES-2655 Simplify JPA RRT javaDoc

These details can be infered via java source code analysis. As no description is added, they are useless.

Not to mention that this generated intelliJ warnings/errors

chibenwa committed:
JAMES-2655 JPA RRT getMapping & getStoredMappings are duplicated

mapAddress can then call getStoredMappings directly

chibenwa committed:
JAMES-2641 Solve concurency issues in event firing test

Tests proved to be unstable. After investigation this is a Reactor related bug.

chibenwa committed:
JAMES-2641 MailboxManagerTest::createMailboxShouldFireMailboxAddedEvent don't require quota feature

chibenwa committed:
Remove James 3.0 roadmap progress

mbaechler committed:
Merge branch 'reactor-blob-store-2'

mbaechler committed:
JAMES-2659 disable instable test

mbaechler committed:
JAMES-2659 remove useless mock library usage

mbaechler committed:
JAMES-2659 add a test to demonstrate that synchronization on listeners is not well done

mbaechler committed:
JAMES-2659 fix both registrationsShouldNotHandleEventsAfterStop tests

mbaechler committed:
JAMES-2659 re-implement guard logic for RabbitEventBus start/stop

aduprat committed:
[maven-release-plugin] prepare release james-project-3.3.0

aduprat committed:
[maven-release-plugin] prepare for next development iteration

aduprat committed:
[maven-release-plugin] Install packages when building integration tests

aduprat committed:
[maven-release-plugin] Spring dockerfile should rely on 3.4.0-SNAPSHOT

Fredmomo committed:
Merge branch 'master' into JAMES-2156

Fredmomo committed:
JAMES-2156 Merge from master and resolve conflicts

Fredmomo committed:
JAMES-2156 FixUp
